src/tokio_reactor.rs

```diff
@@ -0,0 +1,199 @@
+use crate::{
+    Config, Error, MemoryKeychain, Message, Moderator, Notification, Reactor, Request, Response,
+};
+
+use async_trait::async_trait;
+use tokio::sync::mpsc;
+
+use core::time::Duration;
+use std::time::Instant;
+
+/// Communication bridge with a consensus reactor.
+pub struct TokioReactor {
+    timeout: Duration,
+
+    listener: mpsc::Receiver<Message>,
+    sender: mpsc::Sender<Message>,
+
+    /// Reactor will dispatch messages to
+    outbound: mpsc::Sender<Message>,
+}
+
+impl TokioReactor {
+    /// Await for the next message sent from a reactor
+    pub async fn next_async(&mut self) -> Option<Message> {
+        self.listener.recv().await
+    }
+
+    /// Send a notification to the reactor
+    pub async fn notify(&mut self, notification: Notification) -> Result<(), Error> {
+        let notification = Message::Notification(notification);
+
+        self.sender
+            .send_timeout(notification, self.timeout)
+            .await
+            .map_err(|_| Error::ResourceNotAvailable)
+    }
+
+    /// Send a request to the reactor
+    pub async fn request(&mut self, request: Request) -> Result<Response, Error> {
+        let id = request.id();
+        let request = Message::Request(request);
+
+        self.sender
+            .send_timeout(request, self.timeout)
+            .await
+            .map_err(|_| Error::ResourceNotAvailable)?;
+
+        #[cfg(feature = "trace")]
+        tracing::debug!(
+            "request {:?} sent, awaiting response with timeout {:?}",
+            request,
+            self.timeout
+        );
+
+        tokio::time::timeout(self.timeout, self._request(id))
+            .await
+            .map_err(|_e| Error::ResourceNotAvailable)?
+    }
+
+    async fn _request(&mut self, id: u64) -> Result<Response, Error> {
+        loop {
+            match self.listener.recv().await {
+                Some(Message::Response(r)) if r.id() == id => return Ok(r),
+                Some(m) => {
+                    if let Err(_e) = self.outbound.send(m).await {
```

Comment:
Underscore is meant for unused variables
```suggestion
                    if let Err(e) = self.outbound.send(m).await {
```